lib/stm32/common/i2c_common_v1.c

```diff
@@ -463,4 +463,96 @@ void i2c_clear_dma_last_transfer(uint32_t i2c)
 	I2C_CR2(i2c) &= ~I2C_CR2_LAST;
 }
 
+void i2c_set_speed(uint32_t p, enum i2c_speeds speed, uint32_t clock_megahz)
+{
+	i2c_set_clock_frequency(p, clock_megahz);
+	switch(speed) {
+	case i2c_speed_fm_400k:
+		i2c_set_fast_mode(p);
+		i2c_set_ccr(p, clock_megahz * 5 / 6);
+		i2c_set_trise(p, clock_megahz + 1);
+		break;
+	default:
+		/* fall back to standard mode */
+	case i2c_speed_sm_100k:
+		i2c_set_standard_mode(p);
+		/* x Mhz / (100kHz * 2) */
+		i2c_set_ccr(p, clock_megahz * 5);
+		/* Sm mode, (100kHz) freqMhz + 1 */
+		i2c_set_trise(p, clock_megahz + 1);
+		break;
+	}
+}
+
+
+static void i2c_write7_v1(uint32_t i2c, int addr, uint8_t *data, size_t n)
+{
+	while ((I2C_SR2(i2c) & I2C_SR2_BUSY)) {
+	}
+
+	i2c_send_start(i2c);
+
+	/* Wait for master mode selected */
+	while (!((I2C_SR1(i2c) & I2C_SR1_SB)
+		& (I2C_SR2(i2c) & (I2C_SR2_MSL | I2C_SR2_BUSY))));
+
+	i2c_send_7bit_address(i2c, addr, I2C_WRITE);
+
+	/* Waiting for address is transferred. */
+	while (!(I2C_SR1(i2c) & I2C_SR1_ADDR));
+
+	/* Cleaning ADDR condition sequence. */
+	uint32_t reg32 = I2C_SR2(i2c);
+	(void) reg32; /* unused */
+
+	size_t i;
```

Comment:
And the size_t i I'll probably just move into the for loop.  use c99+ as god intended.